utils/src/main/java/software/amazon/awssdk/utils/async/DelegatingBufferingSubscriber.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.utils.async;
+
+import static software.amazon.awssdk.utils.async.StoringSubscriber.EventType.ON_COMPLETE;
+import static software.amazon.awssdk.utils.async.StoringSubscriber.EventType.ON_NEXT;
+
+import java.nio.ByteBuffer;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicLong;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+
+@SdkProtectedApi
+public class DelegatingBufferingSubscriber extends AbstractFlatteningSubscriber<ByteBuffer, ByteBuffer> {
+    /**
+     * The maximum amount of bytes allowed to be stored in the StoringSubscriber
+     */
+    private final long maximumBuffer;
+
+    /**
+     * Current amount of bytes buffered in the StoringSubscriber
+     */
+    private final AtomicLong currentlyBuffered = new AtomicLong(0);
+
+    /**
+     * Stores the bytes received from the upstream publisher, awaiting sending them to the delegate once the buffer size is
+     * reached.
+     */
+    private final StoringSubscriber<ByteBuffer> storage = new StoringSubscriber<>(Integer.MAX_VALUE);
+
+    public DelegatingBufferingSubscriber(long maximumBuffer, Subscriber<? super ByteBuffer> subscriber) {
+        super(subscriber);
+        this.maximumBuffer = maximumBuffer;
+    }
+
+    @Override
+    public void onNext(ByteBuffer item) {
+        if (currentlyBuffered.get() + item.remaining() > maximumBuffer) {
```

Comment:
I think we should trigger a flush as long as we have data instead of buffer is full.